Eliminate the need for inheritance for action controllers.

lbajsarowicz opened the pull request:
### Description (*)
Kindly see issue https://github.com/magento/community-features/issues/9 for detailed information.

I'd like to say great "Thank you" to @Vinai  for letting me work on that.

### Fixed Issues (if relevant)
All the CE unit and integration tests pass. The big question is how the EE will do.
When this PR is merged, frontend action controllers don't have to extend AbstractAction, they only have to implement the ActionInterface.

This PR mainly moves the dispatching of the controller_predispatch events into a plugin.  
It also changes the plugins for `Action::dispatch()` method to `ActionInterface::execute()`.  
Finally, it preserves the behavior in regards to the `NO_DISPATCH` and `NO_POST_DISPATCH` action flags.

Backend action controllers still have to extend the backend module abstract Action for authentication - moving that to a plugin will be a separate PR.

1. https://github.com/magento/magento2/pull/16268
1. https://github.com/magento/magento2/pull/13045

### NOTE:
This PR is a port of #16268 to the 2.4-develop branch

### Related Pull Requests
* https://github.com/magento/magento2/pull/26778
* https://github.com/magento/partners-magento2ee/pull/132
* https://github.com/magento/magento2-page-builder/pull/398
<!-- related pull request placeholder -->

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

Vinai committed:
Move predispatch and postdispatch events from abstract controller to action interface plugin

See https://github.com/magento/community-features/issues/9 for further information.

Vinai committed:
Remove LoD violation which prohibits non inheritance based action controllers

Vinai committed:
Ignore CouplingBetweenObjects in controller action integration test case

Vinai committed:
Use ActionFlag as direct dependency instead of LoD violation

Vinai committed:
Remove void return type to keep code PHP 7.0 compatible

Vinai committed:
Prohibit calling controller execute if the NO_DISPATCHED flag was set

Vinai committed:
Avoid duplicate annotations when it can be used on class

Vinai committed:
Removed unused parameter

Vinai committed:
Move Action::dispatch() plugins to ActionInterface::execute()

Vinai committed:
Move more Action::dispatch() plugins to ActionInterface::execute()

Vinai committed:
Set area for controller tests

Vinai committed:
Add missing @SuppressWarnings(PHPMD.UnusedFormalParameter) to plugin method

Vinai committed:
Remove whitespace at end of line

Vinai committed:
Ignore PHPMD StaticAccess for ObjectManager::getInstance() in test class

Vinai committed:
Remove obsolete plugin method request parameter and clean up test class properties

Vinai committed:
Return void instead of [] from before plugin if no arguments are modified

Vinai committed:
Change the customer notification plugin from AbstractAction:dispatch to ActionInterface::execute

Vinai committed:
Remove training whitespace in empty PHPDoc lines

Vinai committed:
Remove more whitespace at end of line in PHPDoc blocks

Vinai committed:
Add descriptions to the phpdoc blocks, even if they are superfluous

Vinai committed:
Remove comment promising to add the request as a future argument to execute because bc does not allow it

Vinai committed:
Remove training whitespace from comment

Vinai committed:
Fix regression left over from rebase

Vinai committed:
Set area for test so the theme can be loaded

Vinai committed:
Make required changes after cherry-picking commits over from the 2.2-develop branch to the 2.3-develop branch

lbajsarowicz committed:
Merge remote-tracking branch 'origin/2.4-develop' into action-interface

# Conflicts:
#	app/code/Magento/Customer/Model/Plugin/CustomerNotification.php
#	app/code/Magento/Customer/Model/Visitor.php
#	app/code/Magento/Store/etc/di.xml

lbajsarowicz committed:
#16268 Amends to adopted branch

lbajsarowicz committed:
Merge remote-tracking branch 'origin/2.4-develop' into architecture/action-interface

lbajsarowicz committed:
:rofl: debugging on production

lbajsarowicz committed:
Merge remote-tracking branch 'origin/2.4-develop' into architecture/action-interface

lbajsarowicz committed:
Remove temporary debug changes

lbajsarowicz committed:
Code Review changes: Add sortOrder to plugin, avoid returning value from beforeDispatch

lbajsarowicz committed:
Fix static analysis

lbajsarowicz commented:
@magento run all tests

lbajsarowicz commented:
@magento run all tests

lbajsarowicz commented:
@magento run all tests
Rolled back change to Magento EE (getRequest())

lbajsarowicz committed:
Merge remote-tracking branch 'origin/2.4-develop' into architecture/action-interface

lbajsarowicz commented:
With https://github.com/magento/partners-magento2ee/commit/cdcc7030d73716e330602fda19322c9943404817 - there's Integration Test `CRUDTest` introduced that... Introduces some weird behaviour.

@SeruyV Could you explain what actually your test is expected to test and why it collides with ↑?

lbajsarowicz commented:
@magento run all tests

lbajsarowicz commented:
@magento run Functional Tests EE, Integration Tests

ihor-sviziev commented:
@lbajsarowicz right now only test fails that also fails on other PRs, so seems like I can start reviewing this PR 🎆

ihor-sviziev commented on app/code/Magento/Customer/Model/Plugin/CustomerNotification.php:
```diff
@@ -55,39 +61,37 @@ class CustomerNotification
         CustomerRepositoryInterface $customerRepository,
-        LoggerInterface $logger
+        LoggerInterface $logger,
+        RequestInterface $request = null
```
Plugins are not part of `@api` part, so no need to add fallback to object manager here

lbajsarowicz replied:
Nice point

ihor-sviziev commented on app/code/Magento/Customer/Model/Visitor.php:
```diff
@@ -116,7 +129,7 @@ public function __construct(
         $this->dateTime = $dateTime;
         $this->indexerRegistry = $indexerRegistry;
-        $this->requestSafety = $requestSafety ?? ObjectManager::getInstance()->get(RequestSafetyInterface::class);
+        $this->requestSafety = $requestSafety ?? ObjectManager::getInstance()->create(RequestSafetyInterface::class);
```
Why it was replaced to “create”?

lbajsarowicz replied:
Dunno :-) Will change.

ihor-sviziev commented on app/code/Magento/Customer/Model/Visitor.php:
```diff
@@ -17,28 +32,26 @@
     const DEFAULT_ONLINE_MINUTES_INTERVAL = 15;
-
     const XML_PATH_ONLINE_INTERVAL = 'customer/online_customers/online_minutes_interval';
+    const SECONDS_24_HOURS = 86400;
```
```suggestion
    private const SECONDS_24_HOURS = 86400;
```

ihor-sviziev commented on app/code/Magento/Customer/Test/Unit/Model/Plugin/CustomerNotificationTest.php:
```diff
@@ -3,121 +3,126 @@
             ->method('remove')
-            ->with(NotificationStorage::UPDATE_CUSTOMER_SESSION, self::$customerId);
-
-        $this->sessionMock->expects($this->once())->method('setCustomerData')->with($customerMock);
```
Why these checks were removed?

lbajsarowicz replied:
It's not removed, but was introduced in `2.4-develop` in last year :+1:

lbajsarowicz replied:
Actually it's implementation testing, however I'm reverting that change partially.

ihor-sviziev commented on app/code/Magento/Store/etc/di.xml:
```diff
@@ -63,9 +63,11 @@
-        <plugin name="designLoader" type="Magento\Framework\App\Action\Plugin\Design" />
+    <type name="Magento\Framework\App\ActionInterface">
+        <plugin name="storeCheck" type="Magento\Store\App\Action\Plugin\StoreCheck"/>
+        <plugin name="designLoader" type="Magento\Framework\App\Action\Plugin\Design"/>
```
Looks like it should be not in store module, but in di.xml related to framework - global di.xml file

ihor-sviziev commented on app/code/Magento/Store/etc/di.xml:
```diff
@@ -63,9 +63,11 @@
-        <plugin name="designLoader" type="Magento\Framework\App\Action\Plugin\Design" />
+    <type name="Magento\Framework\App\ActionInterface">
+        <plugin name="storeCheck" type="Magento\Store\App\Action\Plugin\StoreCheck"/>
+        <plugin name="designLoader" type="Magento\Framework\App\Action\Plugin\Design"/>
```
Looks like it should be not in store module, but in di.xml related to framework - global di.xml file

lbajsarowicz replied:
Actually - it should not be in `base` scope but `adminhtml`.
And the Plugin is basically related to Backend... So I'll move both Plugin and entry to Backend module.

lbajsarowicz replied:
OK. There's different plugin for Backend. 
That one should be in Store and relate only to Frontend :man_facepalming:

lbajsarowicz replied:
Managed to understand that and doing proper changes.

ihor-sviziev commented:
@lbajsarowicz please also copy PR description from previous PR that you’ve closed.

@lbajsarowicz and @Vinai 
Thank you so much for such amazing contribution!

lbajsarowicz committed:
Merge remote-tracking branch 'origin/2.4-develop' into architecture/action-interface

lbajsarowicz committed:
Code Review changes to Customer module

lbajsarowicz commented:
@magento run all tests

Realised that I introduced duplicate logic to Load Design.
Finally removed the redundant plugin, and moved leftovers to Framework.

lbajsarowicz commented:
D'oh! That was bad move, actually :-(

lbajsarowicz commented:
@magento run all tests

lbajsarowicz commented:
@magento run all tests

ihor-sviziev commented:
@magento give me test instance

magento-engcom-team commented:
Hi @ihor-sviziev. Thank you for your request. I'm working on Magento instance for you

magento-engcom-team commented:
Hi @ihor-sviziev, here is your [new Magento instance](https://pr-26778.instances.magento-community.engineering).
Admin access: [https://pr-26778.instances.magento-community.engineering/admin_4127](https://pr-26778.instances.magento-community.engineering/admin_4127)
Login: `1415070a` Password: `1c5cdeb10854`
Instance will be terminated in up to 3 hours.

ihor-sviziev commented:
Latest changes caused empty product page on frontend, please check it:
![image](https://user-images.githubusercontent.com/1873745/74331900-4ae5cd80-4d9d-11ea-8bca-bdeabe9eeab3.png)

This also caused failures for almost all tests

lbajsarowicz committed:
Code Review changes

lbajsarowicz commented:
@magento run all tests

lbajsarowicz committed:
Fix invalid filename

ihor-sviziev commented:
@magento run all tests

lbajsarowicz committed:
Rollback changes to di.xml entries

lbajsarowicz commented:
@magento run all tests

lbajsarowicz commented:
@ihor-sviziev So it's confirmed.
`di.xml` in Framework is not treated the same way that `di.xml` in module, for some reason :man_facepalming:

ihor-sviziev reviewed (approved):
Failing functional b2b and EE tests looks not related to changes from this PR as fails also on others PRs

magento-engcom-team commented:
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-6882](https://jira.corp.magento.com/browse/ENGCOM-6882) has been created to process this Pull Request

ihor-sviziev commented:
Good job @Vinai @lbajsarowicz!
Thank you so much for such amazing contribution again!

lbajsarowicz commented:
@magento run Functional Tests B2B, Functional Tests EE

lbajsarowicz committed:
Merge remote-tracking branch 'origin/2.4-develop' into architecture/action-interface

lbajsarowicz commented:
@magento run all tests

I've done daily merge of latest `2.4-develop` branches :-(

zbigp commented on app/code/Magento/Store/App/Action/Plugin/StoreCheck.php:
```diff
@@ -1,40 +1,46 @@
      * @return void
      * @SuppressWarnings(PHPMD.UnusedFormalParameter)
-     * @throws \Magento\Framework\Exception\State\InitException
+     * @throws \Magento\Framework\Exception\State\InitExceptionn
```
I think it is a Typo

lbajsarowicz replied:
It actually is.
Will fix this tomorrow.

ihor-sviziev replied:
Fixed by myself

ihor-sviziev commented on app/code/Magento/Store/App/Action/Plugin/StoreCheck.php:
```diff
@@ -1,40 +1,46 @@
      * @return void
      * @SuppressWarnings(PHPMD.UnusedFormalParameter)
-     * @throws \Magento\Framework\Exception\State\InitException
+     * @throws \Magento\Framework\Exception\State\InitExceptionn
```
```suggestion
     * @throws \Magento\Framework\Exception\State\InitException
```

ihor-sviziev committed:
magento/magento2#26778 Eliminate the need for inheritance for action controllers

Fix typo

lbajsarowicz commented:
Thanks

ihor-sviziev commented:
@magento run all tests

lbajsarowicz commented:
@ihor-sviziev 

![giphy](https://user-images.githubusercontent.com/1639941/74726512-e1077100-523f-11ea-8ba7-621b32e6350c.gif)

ihor-sviziev commented:
@lbajsarowicz there is some random functional test failures. They're not related to changes from this PR.
@engcom-Charlie could you get this task to test?

engcom-Alfa commented:
:heavy_check_mark: QA Passed

engcom-Golf commented:
@magento run all tests

Nazar65 committed:
Merge branch '2.4-develop' of https://github.com/magento/magento2 into architecture/action-interface

engcom-Golf commented:
@magento run all tests

engcom-Golf commented:
@magento run all tests

engcom-Golf commented:
@magento run all tests

lbajsarowicz commented:
@ihor-sviziev that's why I'm having fun of it.

sivaschenko commented:
@magento run all tests

lenaorobei commented on lib/internal/Magento/Framework/App/Action/Action.php:
```diff
@@ -17,14 +25,15 @@
  * @SuppressWarnings(PHPMD.CouplingBetweenObjects)
  * @SuppressWarnings(PHPMD.NumberOfChildren)
  */
 abstract class Action extends AbstractAction
```
I believe we need to mark it as deprecated.

lbajsarowicz replied:
It's API

ihor-sviziev replied:
I believe we can depreacate even api 👍

lbajsarowicz replied:
If you say so... :+1:

lbajsarowicz replied:
After adding `@deprecated` it's failing Unit Tests :rofl:

lbajsarowicz committed:
Deprecate Action class

lbajsarowicz committed:
Merge remote-tracking branch 'lbajsarowicz/architecture/action-interface' into architecture/action-interface

lbajsarowicz commented:
@magento run all tests

lbajsarowicz committed:
Merge remote-tracking branch 'origin/2.4-develop' into architecture/action-interface

lbajsarowicz commented:
@magento run all tests

Updated with latest `2.4-develop`, as **Unit Tests** were failing after adding `@deprecated` message :man_facepalming:

lbajsarowicz committed:
Merge remote-tracking branch 'origin/2.4-develop' into architecture/action-interface

lbajsarowicz committed:
Adjust Unit Tests to follow Request fetched on Constructor

lbajsarowicz commented:
@magento run all tests

I found the issue which was related to Unit Test introduced 3 weeks ago.
This was fixed by adding RequestInterface to the constructor instead of being returned by Event Object.

lbajsarowicz commented:
@magento run Functional Tests EE

Tests failed probably because of the change of date.

lbajsarowicz commented:
Yay, we are back on track!
@lenaorobei Can we get an ETA on merge?

lenaorobei commented:
@buskamuza should be able to review this PR from architecture  till the end of Wednesday perspective and then we can move forward.

buskamuza commented on app/code/Magento/Backend/App/Action/Plugin/LoadDesignPlugin.php:
```diff
@@ -0,0 +1,46 @@
+/**
+ * Workaround to load Design before Backend Action dispatch.
+ *
+ * @FIXME Remove when \Magento\Backend\App\AbstractAction::dispatch refactored.
```
Will this be done in scope of this PR?

lbajsarowicz replied:
No, this has to be done in separate PR, as each of these changes makes a huge amount of problems and side effects.

I came with the approach of small steps, instead.

buskamuza replied:
Makes sense.

buskamuza commented on lib/internal/Magento/Framework/App/Action/Action.php:
```diff
@@ -92,35 +102,15 @@ public function dispatch(RequestInterface $request)
-            \Magento\Framework\Profiler::start('action_body');
+            Profiler::start('action_body');
             $result = $this->execute();
-            \Magento\Framework\Profiler::start('postdispatch');
```
Looks like those start/stop profiler actions were lost. Are they removed intentionally to simplify the implementation?

lbajsarowicz replied:
I'm not sure but looks like something I adopted from previous PRs
Let me create another PR based on that one, that introduces `DispatchProfilerPlugin` on `\Magento\Framework\App\FrontControllerInterface::dispatch`

`beforeDispatch` will trigger `Profiler::start`
`afterDispatch` will trigger `Profiler::stop`

Is that fine for you?

lbajsarowicz replied:
OK. Managed to fix that together with Docblock change.

buskamuza reviewed (approved):
Everything looks good, related to the objective of the PR.

One request from my side: please, describe breaking changes. As this is a core area that will touch every page in Magento application, it's important to notify the community about the changes that may potentially affect them. What I noticed on high level:
- though events are preserved (moved to a plugin), object passed to the event is different (it is ActionInterfacde now, and it was AbstractAction before). Most likely nobody will be affected immediately (as all actions had to inherit AbstractAction), but I think it would still make sense to prepare extension developers, so that they can adopt their code for new code.
- also, due to AbstractAction is being replaced with ActionInterface, plugins should be written for ActionInterface now (see changes in this PR - same changes should extension developers do if they pluginize controller actions). I'd suggest at least describe cases that caused us to change plugins

If there are more changes that may affect extension developers, please describe in the PR.

buskamuza commented:
Another big question that bothers me is usage of the Request object...
The PR clearly shows that current design of `ActionInterface` forces other classes to violate Tech Guidelines (`9.10. The Request, Session, and Cookie objects MUST NOT be injected in an object constructor. They MUST be passed only as method arguments.`).
I would not request to fix this in the current PR, but if there is somebody interested, I'd deprecated current `ActionInterface` and introduce new one, which accepts Request object as an argument for `execute()` method. So controllers, plugins, and other presentation layer actors don't have to inject Request in the constructor.

buskamuza commented:
Another big question that bothers me is usage of the Request object...
The PR clearly shows that current design of `ActionInterface` forces other classes to violate Tech Guidelines (`9.10. The Request, Session, and Cookie objects MUST NOT be injected in an object constructor. They MUST be passed only as method arguments.`).
I would not request to fix this in the current PR, but if there is somebody interested, I'd deprecated current `ActionInterface` and introduce new one, which accepts Request object as an argument for `execute()` method. So controllers, plugins, and other presentation layer actors don't have to inject Request in the constructor.

lbajsarowicz commented:
@buskamuza I like that you are pointing that out. We also discussed some of the architectural concerns with @Vinai during our session, but we decided to make little steps, instead of doing everything with a single step.

lenaorobei commented:
@buskamuza thank you for the review. I created a separate task to cover the documentation piece https://github.com/magento/devdocs/issues/6721

lenaorobei commented on lib/internal/Magento/Framework/App/Action/Action.php:
```diff
@@ -3,28 +3,38 @@
  * Action classes that do not extend from this class will lose this behavior and might not function correctly
  *
- * TODO: Remove this class. Allow implementation of Action Controllers by just implementing Action Interface.
+ * @TODO: Remove this class. Allow implementation of Action Controllers by just implementing Action Interface.
```
This message is a little confusing. Thanks to this PR we will `Allow implementation of Action Controllers by just implementing Action Interface.`. Let's just remove it. `@deprecated`  should be good enough.

lbajsarowicz replied:
That is fine. I am going to remove `@TODO`, also I would like to change wording for `@deprecated`
```php
* @deprecated 2.4.0, use \Magento\Framework\App\ActionInterface
```

lbajsarowicz committed:
Restore Profiler for `postdispatch`, remove `@TODO` and change `@deprecated` message

lbajsarowicz commented:
@magento run all tests

Changes introduced to `magento/magento2` and `magento/partners-magento2ee`

lenaorobei committed:
Merge branch '2.4-develop' into architecture/action-interface

lenaorobei commented:
@magento run all tests

magento-engcom-team commented:
Hi @lenaorobei, thank you for the review.
[ENGCOM-6882](https://jira.corp.magento.com/browse/ENGCOM-6882) has been created to process this Pull Request

lbajsarowicz commented:
@magento run all tests

Integration Tests of PageBuilder failed due to missing `@magentoAppArea adminhtml`.
I also improved the Integration Test to follow part of _Good Practices_

lbajsarowicz commented:
@slavvka Could you re-run tests when we clean up some space on Testing Environment?
![image](https://user-images.githubusercontent.com/1639941/75597451-d053c800-5a95-11ea-84f5-0809ed02cccc.png)
